acvm-repo/acvm/tests/solver.rs

```diff
@@ -722,3 +724,531 @@ fn memory_operations() {
+        255, 255, 255, 255, 255, 0, 0, 0, 0, 255, 255, 255, 255,
+    ];
+
+    vec![bn254_fq, bn254_fr, secpk1_fr, secpk1_fq, secpr1_fq, secpr1_fr]
+}
+
+prop_compose! {
+    fn bigint_with_modulus()(modulus in select(allowed_bigint_moduli()))
+        (input in proptest::collection::vec(any::<(u8, bool)>(), modulus.len()), modulus in Just(modulus))
+        -> (Vec<(FieldElement, bool)>, Vec<u8>) {
+        let input = input.into_iter().map(|(x, use_constant)| {
+            (FieldElement::from(x as u128), use_constant)
+        }).collect();
+        (input, modulus)
+    }
+}
+
+prop_compose! {
+    fn bigint_pair_with_modulus()(input_modulus in bigint_with_modulus())
+        (ys in proptest::collection::vec(any::<(u8, bool)>(), input_modulus.1.len()), input_modulus in Just(input_modulus))
+        -> (Vec<(FieldElement, bool)>, Vec<(FieldElement, bool)>, Vec<u8>) {
+        let ys = ys.into_iter().map(|(x, use_constant)| {
+            (FieldElement::from(x as u128), use_constant)
+        }).collect();
+        (input_modulus.0, ys, input_modulus.1)
+    }
+}
+
+prop_compose! {
+    fn bigint_triple_with_modulus()(xs_ys_modulus in bigint_pair_with_modulus())
+        (zs in proptest::collection::vec(any::<(u8, bool)>(), xs_ys_modulus.2.len()), xs_ys_modulus in Just(xs_ys_modulus))
+        -> (Vec<(FieldElement, bool)>, Vec<(FieldElement, bool)>, Vec<(FieldElement, bool)>, Vec<u8>) {
+        let zs = zs.into_iter().map(|(x, use_constant)| {
+            (FieldElement::from(x as u128), use_constant)
+        }).collect();
+        (xs_ys_modulus.0, xs_ys_modulus.1, zs, xs_ys_modulus.2)
+    }
+}
+
+fn bigint_add_op() -> BlackBoxFuncCall<FieldElement> {
+    BlackBoxFuncCall::BigIntAdd { lhs: 0, rhs: 1, output: 2 }
+}
+
+fn bigint_mul_op() -> BlackBoxFuncCall<FieldElement> {
+    BlackBoxFuncCall::BigIntMul { lhs: 0, rhs: 1, output: 2 }
+}
+
+fn bigint_sub_op() -> BlackBoxFuncCall<FieldElement> {
+    BlackBoxFuncCall::BigIntSub { lhs: 0, rhs: 1, output: 2 }
+}
+
+fn bigint_div_op() -> BlackBoxFuncCall<FieldElement> {
+    BlackBoxFuncCall::BigIntDiv { lhs: 0, rhs: 1, output: 2 }
+}
+
+// Input is a BigInt, represented as a LE Vec of u8-range FieldElement's along
+// with their use_constant values.
+//
+// Output is a zeroed BigInt with the same byte-length and use_constant values
+// as the input.
+fn bigint_zeroed(input: &[(FieldElement, bool)]) -> Vec<(FieldElement, bool)> {
+    input.iter().map(|(_, use_constant)| (FieldElement::zero(), *use_constant)).collect()
+}
+
+// bigint_zeroed, but returns one
+fn bigint_to_one(input: &[(FieldElement, bool)]) -> Vec<(FieldElement, bool)> {
+    let mut one = bigint_zeroed(input);
+    // little-endian
+    one[0] = (FieldElement::one(), one[0].1);
+    one
+}
+
+fn drop_use_constant(input: &[(FieldElement, bool)]) -> Vec<FieldElement> {
+    input.iter().map(|x| x.0).collect()
+}
+
+fn bigint_solve_binary_op_opt(
```

Comment:
it's not clear to me what this function is doing, and I see some duplicate codes, could you refactor it with comments and maybe some helper functions?